numpy/core/tests/test_ufunc.py

```diff
@@ -1936,3 +1936,38 @@ def test_nat_is_not_inf(self, nat):
             assert not np.isinf(nat)
         except TypeError:
             pass  # ok, just not implemented
+
+
+@pytest.mark.parametrize('ufunc', [getattr(np, x) for x in dir(np)
+                                if isinstance(getattr(np, x), np.ufunc)])
+def test_ufunc_types(ufunc):
+    '''
+    Check all ufuncs that the correct type is returned. Avoid
+    object and boolean types since many operations are not defined for
+    for them.
+
+    Choose the shape so even dot and matmul will succeed
+    '''
+    for typ in ufunc.types:
+        # types is a list of strings like ii->i
+        if 'O' in typ or '?' in typ:
+            continue
+        inp, out = typ.split('->')
+        if ',' in inp:
+            args = [np.zeros((3, 3), t.strip()) for t in inp.split(',')]
+        else:
+            args = [np.ones((3, 3), t.strip()) for t in inp]
+        outt = out.strip()
```

Comment:
What if there are multiple outputs, possibly of different types? There are 23 like that.

Reply:
I see this is handled below.